cfme/infrastructure/virtual_machines.py

```diff
@@ -86,46 +87,169 @@
 retire_remove_button = "//span[@id='remove_button']/a/img"
 
 match_page = partial(match_location, controller='vm_infra', title='Virtual Machines')
-vm_templates_tree = partial(accordion.tree, "VMs & Templates")
-vms_tree = partial(accordion.tree, "VMs")
-templates_tree = partial(accordion.tree, "Templates")
-
-
-def reset_page():
-    tb.select("Grid View")
-    from cfme.web_ui import paginator
-    if sel.is_displayed(search_box.search_field):
-        search.ensure_normal_search_empty()
-    if paginator.page_controls_exist():
-        # paginator.results_per_page(1000)
-        paginator.check_all()
-        paginator.uncheck_all()
 
 
 drift_table = CheckboxTable("//th[normalize-space(.)='Timestamp']/ancestor::table[1]")
 
 
-class InfraVmDetailsToolbar(View):
-    configuration = Dropdown("Configuration")
-    policy = Dropdown("Policy")
-    lifecycle = Dropdown("Lifecycle")
+class InfraVmTemplatesGenericDetailsToolbar(View):
+    reload = Button(title='Reload current display')
+    configuration = Dropdown('Configuration')
+    policy = Dropdown('Policy')
     monitoring = Dropdown("Monitoring")
-    power = Dropdown("Power")
+    download = Button(title='Download summary in PDF format')
+    lifecycle = Dropdown('Lifecycle')
+
+
+class InfraVmDetailsToolbar(InfraVmTemplatesGenericDetailsToolbar):
+    """Toolbar for VM details differs from All VMs&TemplatesView
+    """
     access = Dropdown("Access")
+    power = Dropdown('VM Power Functions')
+
 
+class VmsTemplatesAccordion(View):
+    """
+    The accordion on the Virtual Machines page
+    """
+    @View.nested
+    class vmstemplates(Accordion):  # noqa
+        ACCORDION_NAME = 'VMs & Templates'
+        tree = ManageIQTree()
+
+    @View.nested
+    class vms(Accordion):  # noqa
+        ACCORDION_NAME = 'VMs'
+        tree = ManageIQTree()
 
-class InfraVmDetailsView(BaseLoggedInPage):
-    # TODO this is only minimal implementation for toolbar access through widgetastic
-    toolbar = View.nested(InfraVmDetailsToolbar)
+    @View.nested
+    class templates(Accordion):  # noqa
+        ACCORDION_NAME = 'Templates'
+        tree = ManageIQTree()
+
+
+class InfraVMView(BaseLoggedInPage):
+    """Base view for header/nav check, inherit for navigatable views"""
+
+    @property
+    def in_infra_vms(self):
+        return (
+            self.logged_in_as_current_user and
+            self.navigation.currently_selected == ['Compute', 'Infrastructure',
+                                                   'Virtual Machines'] and
+            match_location(controller='vm_infra', title='Virtual Machines'))
+
+
+class VmsTemplatesAllView(InfraVMView):
+    """
+    The collection page for instances
+    """
+    @property
+    def is_displayed(self):
+        return (
+            self.in_infra_vms and
+            self.sidebar.vmstemplates.tree.currently_selected == 'All VMs & Templates' and
+            self.entities.title.text == 'All VMs & Templates')
```

Comment:
Put is_displayed after widget definitions, this does not look good :)

Reply:
(applies for all such occurences)

Reply:
done